### PR TITLE
[SPARK-15674][SQL] Deprecates "CREATE TEMPORARY TABLE USING...", uses "CREAT TEMPORARY VIEW USING..." instead

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

The current implementation of "CREATE TEMPORARY TABLE USING datasource..." is NOT creating any intermediate temporary data directory like temporary HDFS folder, instead, it only stores a SQL string in memory. Probably we should use "TEMPORARY VIEW" instead.

This PR assumes a temporary table has to link with some temporary intermediate data. It follows the definition of temporary table like this (from [hortonworks doc](https://docs.hortonworks.com/HDPDocuments/HDP2/HDP-2.3.0/bk_dataintegration/content/temp-tables.html)):

> A temporary table is a convenient way for an application to automatically manage intermediate data generated during a complex query

**Example**:

```
scala> spark.sql("CREATE temporary view  my_tab7 (c1: String, c2: String)  USING org.apache.spark.sql.execution.datasources.csv.CSVFileFormat  OPTIONS (PATH '/Users/seanzhong/csv/cars.csv')")
scala> spark.sql("select c1, c2 from my_tab7").show()
+----+-----+
|  c1|   c2|
+----+-----+
|year| make|
|2012|Tesla|
...
```

It NOW prints a **deprecation warning** if "CREATE TEMPORARY TABLE USING..." is used.

```
scala> spark.sql("CREATE temporary table  my_tab7 (c1: String, c2: String)  USING org.apache.spark.sql.execution.datasources.csv.CSVFileFormat  OPTIONS (PATH '/Users/seanzhong/csv/cars.csv')")
16/05/31 10:39:27 WARN SparkStrategies$DDLStrategy: CREATE TEMPORARY TABLE tableName USING... is deprecated, please use CREATE TEMPORARY VIEW viewName USING... instead
```
## How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by hvanhovell** [comment]
So I am not sure I understand this one. Why should we deprecate this in favour of creating a view? A `create temp table ... using` statement describes the access to a physical storage; which in my book is a table.

Could you elaborate on why we need this?

**Comment by clockfly** [comment]
@hvanhovell 

> create temp table ... using statement describes the access to a physical storage; which in my book is a table.

We still allow `create table using...`, what we deprecate is "create **temporary** table using...". We still treats external data source as tables. To wrap the external data source as a table, we can use "create table using..." .

temporary views and temporary tables are intermediate layers between user and the actual table:

```
User --> Temporary view/table --> External data source table (Can be wrapped by create table using...)
```

Currently, in our implementation, we don't support temporary table, we only supports temporary view. The difference is that:
1. Temporary view is backed by a SQL string, which acts like a pointer. So every time we use the temporary view, the SQL is RE-executed again, which will asks data from the original data source.
2. Temporary table is supposed to execute the SQL for ONLY once, and store the result in a temporary HDFS directory. Every time you use the temporary table, we are actually using the data in the temporary HDFS directory directly, without bothering the original data source.

**Comment by clockfly** [comment]
@hvanhovell 

I updated the description, please check whether it makes more sense now.

**Comment by hvanhovell** [comment]
@clockfly the description is getting there. IIUC the problem we are solving is the following:

`CREATE TEMPORARY TABLE ... USING ...` allows us to create a temporary (session bound) connection to a (potentially) permanent data store. When the session finishes, the table **definition** (connection) is dropped, but the **data** is not. This is more-or-less the behavior you expect with a `TEMPORARY EXTERNAL` table (do we have those?), and this actually violates the common definition of a temporary table in which both the table definition and the data are session bound.

Using `CREATE TEMPORARY VIEW ... USING ...` accomplishes two things:
- It doesn't make assumption about the underlying data (it can both be permanent or session bound).
- It doesn't allow user to write to the datasource.

I do have a couple of issues with this:
- Using `CREATE TEMPORARY TABLE ... USING ...` should still be allowed to use if you are using an actual session local temporary table. We could detect these by checking if a schema is defined (the location is also an issue). How do we deal with this use case?
- I would support creating a `CREATE TEMPORARY EXTERNAL TABLE ... USING ...` to retain the current behavior.

What do you think?

**Comment by clockfly** [comment]
@hvanhovell Probably we can talk more face to face next week.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -90,6 +90,8 @@ statement
         identifierCommentList? (COMMENT STRING)?
         (PARTITIONED ON identifierList)?
         (TBLPROPERTIES tablePropertyList)? AS query                    #createView
+    | CREATE (OR REPLACE)? TEMPORARY VIEW tableIdentifier ('(' colTypeList ')')? tableProvider
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: NIT: Could you break this line up so we keep all `#...` hooks on the same column...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -344,6 +344,19 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
   }
 
   /**
+   * Creates a [[CreateTempViewUsing]] logical plan.
+   */
+  override def visitCreateTempViewUsing(ctx: CreateTempViewUsingContext)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: NIT: Style move the `ctx` to the next line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 3]
```diff
@@ -344,6 +344,19 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+    : LogicalPlan = withOrigin(ctx) {
+    val options = Option(ctx.tablePropertyList).map(visitPropertyKeyValues).getOrElse(Map.empty)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: NIT: you could create the `CreateTempViewUsing` immediately

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 4]
```diff
@@ -375,9 +375,12 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
   object DDLStrategy extends Strategy {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
       case c: CreateTableUsing if c.temporary && !c.allowExisting =>
+        logWarning(
+          s"CREATE TEMPORARY TABLE tableName USING... is deprecated, please use CREATE " +
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: `tableName` is not interpolated at the moment.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 5]
```diff
@@ -115,6 +115,10 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
     if (!path.endsWith(File.separator)) path + File.separator else path
   }
 
+  private def testFile(fileName: String): String = {
+    Thread.currentThread().getContextClassLoader.getResource(fileName).toString
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Why is this needed?

**Comment by hvanhovell** [comment]
@clockfly this looks pretty good. I have left some (minor) comments.

**Comment by clockfly** [comment]
@hvanhovell  Thanks for the review. 

Updated.

**Comment by hvanhovell** [comment]
LGTM pending Jenkins

**Comment by hvanhovell** [comment]
Thanks! Merging to master/2.0
